cpp/src/text/subword/bpe_tokenizer.cu

```diff
@@ -0,0 +1,553 @@
+ * @endcode
+ */
+struct build_encoding_fn {
+  cudf::column_device_view const d_strings;
+  cudf::size_type const* d_byte_indices;
+  cudf::offset_type const* d_offsets;
+  char* d_chars{};
+
+  __device__ void operator()(cudf::size_type idx)
+  {
+    if (d_strings.is_null(idx)) { return; }
+    auto const d_str = get_first_token(d_strings.element<cudf::string_view>(idx));
+    if (d_str.empty()) { return; }
+
+    auto const offset = d_strings.child(cudf::strings_column_view::offsets_column_index)
+                          .element<cudf::offset_type>(idx);
+    auto const d_indices = d_byte_indices + offset;
+    auto d_output        = d_chars ? d_chars + d_offsets[idx] : nullptr;
+
+    // copy chars while indices[i]==0,
+    // insert space each time indices[i]!=0
+    auto const begin = d_indices;
+    auto const end   = d_indices + d_str.size_bytes();
+    auto d_input     = d_str.data();
+    *d_output++      = *d_input++;
+    auto itr         = begin + 1;
+    while (itr < end) {
+      if (*itr++) *d_output++ = ' ';
+      *d_output++ = *d_input++;
+    }
+    // https://github.com/rapidsai/cudf/pull/10270/files#r826319405
+  }
+};
+
+/**
+ * @brief Perform byte pair encoding on each string in the input column.
+ *
+ * The result is a strings column of the same size where each string has been encoded.
+ *
+ * The encoding is performed iteratively. Each pass determines the string's lowest
+ * ranked merge pair as determined by the strings in `merges_table`. This pair
+ * is removed (virtually) from each string before starting the next iteration.
+ *
+ * Once all pairs have exhausted for all strings, the output is constructed from
+ * the results by adding spaces between each remaining pair in each string.
+ *
+ * @param input Strings to encode.
+ * @param merge_pairs Merge pairs data and map used for encoding.
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ */
+std::unique_ptr<cudf::column> byte_pair_encoding(
+  cudf::strings_column_view const& input,
+  bpe_merge_pairs::bpe_merge_pairs_impl const& merge_pairs,
+  rmm::cuda_stream_view stream)
+{
+  CUDF_EXPECTS(!merge_pairs.get_merge_pairs().is_empty(), "Merge pairs table must not be empty");
+
+  // build working vector to hold index values per byte
+  rmm::device_uvector<cudf::size_type> d_byte_indices(input.chars().size(), stream);
+
+  auto const d_merges  = cudf::column_device_view::create(merge_pairs.get_merge_pairs(), stream);
+  auto const d_strings = cudf::column_device_view::create(input.parent(), stream);
+
+  auto offsets   = cudf::make_numeric_column(cudf::data_type{cudf::type_to_id<cudf::offset_type>()},
+                                           static_cast<cudf::size_type>(input.size() + 1),
+                                           cudf::mask_state::UNALLOCATED,
+                                           stream,
+                                           rmm::mr::get_current_device_resource());
```

Comment:
Does this function need to accept an `mr` since it returns memory allocated within the function? It looks like this memory ends up being returned in final result, which is a list column view. The calling function uses an `mr` parameter for other allocations.

Reply:
This is an internal (unnamed namespace function) and the resulting object (offsets) is never returned to the API caller.

Reply:
I thought `return make_strings_column(input.size(), std::move(offsets), std::move(chars), 0, rmm::device_buffer{});` would use move semantics to return a string column with a child column of offsets allocated with this resource. I might be missing something?

Reply:
Yes, the returned strings column is also not returned to the caller of the external API.
The [offsets](https://github.com/rapidsai/cudf/pull/10270/files#diff-21153977ac1ba7735c5cadd16b3c6f1ad6accee8cdc2e30bcb9ae2c27fcdfc59R351) are used to build a [strings column](https://github.com/rapidsai/cudf/pull/10270/files#diff-21153977ac1ba7735c5cadd16b3c6f1ad6accee8cdc2e30bcb9ae2c27fcdfc59R382) which is then used to build a [lists column view](https://github.com/rapidsai/cudf/pull/10270/files#diff-21153977ac1ba7735c5cadd16b3c6f1ad6accee8cdc2e30bcb9ae2c27fcdfc59R525) which is finally used to build an [output strings column](https://github.com/rapidsai/cudf/pull/10270/files#diff-21153977ac1ba7735c5cadd16b3c6f1ad6accee8cdc2e30bcb9ae2c27fcdfc59R534). Only building the output strings column requires the passed in `mr`.

Reply:
Aha! I didn’t fully understand the last step until now. Thank you for your insight!